### PR TITLE
Add support for Azure function deployment.

### PR DESCRIPTION
**Pull request by praveenkumarct**
- Add Azure function app deployment support for both extension and notification modules.
- Documentation for the same.

Addresses:

Issue: [#825](https://github.com/commercetools/commercetools-adyen-integration/issues/825)
PR: https://github.com/commercetools/commercetools-adyen-integration/pull/885

### CODE REVIEW

File: notification/index.azureFunction.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import notificationHandler from './src/handler/notification/notification.handler.js'
+import { getLogger } from './src/utils/logger.js'
+import utils from './src/utils/commons.js'
+import { getErrorCause, isRecoverableError } from './src/utils/error-utils.js'
+import { getCtpProjectConfig, getAdyenConfig } from './src/utils/parser.js'
+
+const logger = getLogger()
+
+export const azureExtensionTrigger = async function (context, event) {
+  const { notificationItems } = event.body
```
Comment (leungkinghin) [inline notification/index.azureFunction.js:10]: what if `body` object is absent in `event`?

**Review by leungkinghin — CHANGES_REQUESTED** [review]
Also, would you add unit test for azure function? I found we have unit test for google function and aws lambda

File: notification/index.azureFunction.js (hunk 1 above)
Comment (praveenkumarct) [reply notification/index.azureFunction.js:10]: Then we don't have notificationItems, So it will return `No notification received`.

Comment (praveenkumarct) [reply notification/index.azureFunction.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/e980819c0531bad932f8ab1bf65462a86dddf7fc

I added an additional check.

**Review by leungkinghin — APPROVED** [review]

File: deployment-examples/azure/serverless/extension/README.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+## Deploy as Azure function app
+
+1. In order to setup the extension module config, follow our [how to run guide](../../../../extension/docs/HowToRun.md).
+2. Extension module requires `ADYEN_INTEGRATION_CONFIG` as `environment variable` or `.extensionrc` file to configure with [external file configuration](../../../../extension/docs/HowToRun.md#external-file-configuration).
+3. Deployments work by uploading an archive containing your function's source code and their dependencies. 
+You could use a deployment package to deploy your function code to Azure function app.
+To bundle the extension module, navigate into `extension` folder and run `npm run zip-azure-function`. 
+4. By default, the Functions runtime looks for your function in index.js.
+    > In the default case, your exported function should be the only export from its file or the export named run or index.
+5. Please check our [faq](../../../../docs/FAQ.md) for your deployment.
```
Comment (lojzatran) [inline deployment-examples/azure/serverless/extension/README.md:10]:
```suggestion
5. Please check our [FAQ](../../../../docs/FAQ.md) for your deployment.
```

File: extension/index.azureFunction.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        {
+          code: 'InvalidInput',
+          message: 'Invalid request body.',
+        },
+      ]
+      handleErrorResponse(context, errors)
+      return
+    }
+
+    const authToken = getAuthorizationRequestHeader(req)
```
Comment (lojzatran) [inline extension/index.azureFunction.js:40]:
```suggestion
  const paymentObj = req.body?.resource?.obj
  if (!paymentObj) {
    const errors = [
      {
        code: 'InvalidInput',
        message: 'Invalid request body.',
      },
    ]
    handleErrorResponse(context, errors)
    return
  }

  const authToken = getAuthorizationRequestHeader(req)
  try {
```

File: extension/index.azureFunction.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import utils from './src/utils.js'
+import paymentHandler from './src/paymentHandler/payment-handler.js'
+import { getAuthorizationRequestHeader } from './src/validator/authentication.js'
+
+const FAILED_VALIDATION = 'FailedValidation'
+
+function handleSuccessResponse(context, paymentResult) {
```
Comment (lojzatran) [inline extension/index.azureFunction.js:7]: Looks like it's not always Success response?

File: notification/index.azureFunction.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      'Unexpected exception occurred.'
+    )
+
+    if (isRecoverableError(err)) {
+      handleErrorResponse(context, 500, cause.message)
+      return
+    }
+  }
+
+  context.res = {
```
Comment (lojzatran) [inline notification/index.azureFunction.js:52]: Can you create `handleXXXResponse` so it's consistent like the other azure class in extension?

File: deployment-examples/azure/serverless/notification/README.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
+## Deploy as Azure function app
+
+1. In order to setup the notification module config, follow our [how to run guide](../../../../notification/docs/HowToRun.md).
+2. Extension module requires `ADYEN_INTEGRATION_CONFIG` as `environment variable` or `.notificationrc` file to configure with [external file configuration](../../../../notification/docs/HowToRun.md#external-file-configuration).
+3. Deployments work by uploading an archive containing your function's source code and their dependencies. 
+You could use a deployment package to deploy your function code to Azure function app.
+To bundle the notification module, navigate into `notification` folder and run `npm run zip-azure-function`. 
+4. By default, the Functions runtime looks for your function in index.js.
+    > In the default case, your exported function should be the only export from its file or the export named run or index.
+5. Please check our [faq](../../../../docs/FAQ.md) for your deployment.
```
Comment (lojzatran) [inline deployment-examples/azure/serverless/notification/README.md:10]:
```suggestion
5. Please check our [FAQ](../../../../docs/FAQ.md) for your deployment.
```

File: deployment-examples/azure/serverless/extension/README.md [hunk 7]
```diff
@@ -0,0 +1,15 @@
+## Deploy as Azure function app
+
+1. In order to setup the extension module config, follow our [how to run guide](../../../../extension/docs/HowToRun.md).
+2. Extension module requires `ADYEN_INTEGRATION_CONFIG` as `environment variable` or `.extensionrc` file to configure with [external file configuration](../../../../extension/docs/HowToRun.md#external-file-configuration).
```
Comment (lojzatran) [inline deployment-examples/azure/serverless/extension/README.md:4]: What is the reason of the file `.extensionrc` for azure?

File: deployment-examples/azure/serverless/notification/README.md [hunk 8]
```diff
@@ -0,0 +1,15 @@
+## Deploy as Azure function app
+
+1. In order to setup the notification module config, follow our [how to run guide](../../../../notification/docs/HowToRun.md).
+2. Extension module requires `ADYEN_INTEGRATION_CONFIG` as `environment variable` or `.notificationrc` file to configure with [external file configuration](../../../../notification/docs/HowToRun.md#external-file-configuration).
```
Comment (lojzatran) [inline deployment-examples/azure/serverless/notification/README.md:4]: What's the reason for `.notificationrc`?

Comment (praveenkumarct) [reply deployment-examples/azure/serverless/notification/README.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/9a641f16cdbcf60dbd91701cc93d098faa79e0a1

File: deployment-examples/azure/serverless/extension/README.md (hunk 7 above)
Comment (praveenkumarct) [reply deployment-examples/azure/serverless/extension/README.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/9a641f16cdbcf60dbd91701cc93d098faa79e0a1

File: extension/index.azureFunction.js (hunk 3 above)
Comment (praveenkumarct) [reply extension/index.azureFunction.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/4cd805a7e68ef7bc6f541c5098ad33b981332a2d

File: deployment-examples/azure/serverless/notification/README.md (hunk 8 above)
Comment (praveenkumarct) [reply deployment-examples/azure/serverless/notification/README.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/addec52897da5c24a72102e9b6438dfcd40ca2d4

Removed rc file usage in Azure and AWS docs.

File: deployment-examples/azure/serverless/extension/README.md (hunk 7 above)
Comment (praveenkumarct) [reply deployment-examples/azure/serverless/extension/README.md]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/addec52897da5c24a72102e9b6438dfcd40ca2d4

Removed rc file usage in Azure and AWS docs.

File: notification/index.azureFunction.js (hunk 5 above)
Comment (praveenkumarct) [reply notification/index.azureFunction.js:52]: I don't see any benefit here. In extension, we have success response called more than once, So I have extracted common method.

File: extension/index.azureFunction.js (hunk 4 above)
Comment (praveenkumarct) [reply extension/index.azureFunction.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/01d0148e062aa412679216055d8cf4ac9d78c53e

File: notification/index.azureFunction.js (hunk 5 above)
Comment (lojzatran) [reply notification/index.azureFunction.js:52]: > I don't see any benefit here. In extension, we have success response called more than once, So I have extracted common method.

OK, but you know that in extension you also have success response only once :)

File: extension/index.azureFunction.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
+import utils from './src/utils.js'
+import paymentHandler from './src/paymentHandler/payment-handler.js'
+import { getAuthorizationRequestHeader } from './src/validator/authentication.js'
+
+const FAILED_VALIDATION = 'FailedValidation'
+
+function handleResponse(context, paymentResult) {
```
Comment (lojzatran) [inline extension/index.azureFunction.js:7]: Currently you have handleResponse that actually handles also error responses and you have `handleErrorResponse` separately. 
Better use this method only for `handleSuccessResponse`. Then you would have `handleSuccessResponse` and `handleErrorResponse` which makes more sense.

The decision which response it is should be in the method `azureExtensionTrigger` like already have now. It doesn't make sense to have this decision both in `azureExtensionTrigger` and in `handleResponse`. So sth like this in `azureExtensionTrigger`:
```
if (paymentResult.actions)
    handleSuccessResponse()
else
    handleErrorResponse()
```

Comment (praveenkumarct) [reply extension/index.azureFunction.js]: https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/b585915c2e020a647de58ae5551a0f52c997fd95

**Review by lojzatran — APPROVED** [review]

**Comment by praveenkumarct** [comment]
I have added trigger function and some settings(host.json, function.json) to both extension and notification modules.
@lojzatran @leungkinghin can you check it one more time and review it.

**Comment by lojzatran** [comment]
> I have added trigger function and some settings(host.json, function.json) to both extension and notification modules. @lojzatran @leungkinghin can you check it one more time and review it.

Thanks for notifying about the files you added. Can you also add to the docs why you need these files? I do not see you use them anywhere nor they are described in any README files.

**Comment by praveenkumarct** [comment]
https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/c21843bc8ccb60f1b025fea1dc28f30b876d114e
Updated docs to inform user the location of Function.json and Host.json.
Also provided links where they can learn more about these settings.

**Comment by lojzatran** [comment]
> [c21843b](https://github.com/commercetools/commercetools-adyen-integration/commit/c21843bc8ccb60f1b025fea1dc28f30b876d114e) Updated docs to inform user the location of Function.json and Host.json. Also provided links where they can learn more about these settings.

Looks good. There is one file `local.settings.json` that is not mentioned. What is that? How can I use this file?

**Comment by praveenkumarct** [comment]
> > [c21843b](https://github.com/commercetools/commercetools-adyen-integration/commit/c21843bc8ccb60f1b025fea1dc28f30b876d114e) Updated docs to inform user the location of Function.json and Host.json. Also provided links where they can learn more about these settings.
> 
> Looks good. There is one file `local.settings.json` that is not mentioned. What is that? How can I use this file?

It is just an additional file, provided for local development if required. But let us not confuse users or developers by adding about this in the docs. We can even delete this file. But I think it won't harm us in anyway, since it is already added in the gitignore file.

**Comment by lojzatran** [comment]
> > > [c21843b](https://github.com/commercetools/commercetools-adyen-integration/commit/c21843bc8ccb60f1b025fea1dc28f30b876d114e) Updated docs to inform user the location of Function.json and Host.json. Also provided links where they can learn more about these settings.
> > 
> > 
> > Looks good. There is one file `local.settings.json` that is not mentioned. What is that? How can I use this file?
> 
> It is just an additional file, provided for local development if required. But let us not confuse users or developers by adding about this in the docs. We can even delete this file. But I think it won't harm us in anyway, since it is already added in the gitignore file.

It's confusing for me why is it there and I am a developer 😄 So you have to clarify it or remove it if it's useless. You can add a section like Local development for azure or sth like that and mention this file there.

**Comment by praveenkumarct** [comment]
https://github.com/commercetools/commercetools-adyen-integration/pull/993/commits/ecdc956f686dab3235217c79d8cadaf87af5fcda

done, No more confusion :)

**Review by lojzatran — APPROVED** [review]
It would be great if you could revert changes for .gitignore file.

**Pull request merged**